### PR TITLE
[Ingest] Add meta processor that allows to modify the metadata attributes of document being processed

### PR DESCRIPTION
**Pull request by martijnvg**
This also adds template support only for the new `meta` processor. I opened #14990 to discuss how template support should further be adopted in ingest.

PR for #14644

### CODE REVIEW

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java [hunk 1]
```diff
@@ -40,22 +42,46 @@ public PipelineExecutionService(PipelineStore store, ThreadPool threadPool) {
...
+                if (ingestDocument.isMetaDataModified()) {
+                    indexRequest.index(ingestDocument.getMetadata(IngestDocument.MetaData.INDEX));
+                    indexRequest.type(ingestDocument.getMetadata(IngestDocument.MetaData.TYPE));
+                    indexRequest.id(ingestDocument.getMetadata(IngestDocument.MetaData.ID));
+                    indexRequest.routing(ingestDocument.getMetadata(IngestDocument.MetaData.ROUTING));
+                    indexRequest.parent(ingestDocument.getMetadata(IngestDocument.MetaData.PARENT));
+                    indexRequest.timestamp(ingestDocument.getMetadata(IngestDocument.MetaData.TIMESTAMP));
+                    String ttlStr = ingestDocument.getMetadata(IngestDocument.MetaData.TTL);
+                    if (ttlStr != null) {
+                        indexRequest.ttl(Long.parseLong(ttlStr));
```
Comment (javanna) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java]: should we validate that the ttl is a valid long as part of the metadata processor? We could add a small validate method to the enum, which does nothing by default but that the ttl one implements.

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java [hunk 2]
```diff
@@ -40,22 +42,46 @@ public PipelineExecutionService(PipelineStore store, ThreadPool threadPool) {
...
+                ttl = String.valueOf(indexRequest.ttl());
+            }
+            Map<String, Object> sourceAsMap = indexRequest.sourceAsMap();
+            IngestDocument ingestDocument = new IngestDocument(index, type, id, routing, parent, timestamp, ttl, sourceAsMap);
+            try {
+                pipeline.execute(ingestDocument);
+                if (ingestDocument.isSourceModified()) {
+                    indexRequest.source(ingestDocument.getSource());
+                }
+                if (ingestDocument.isMetaDataModified()) {
```
Comment (javanna) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java]: I think see the point of not setting the source if it was not modified, but when it comes to metadata, should we just set them back all the time? anyway we end up with a single flag for all of them...

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/WriteableIngestDocument.java [hunk 3]
```diff
@@ -67,13 +69,17 @@ public void writeTo(StreamOutput out) throws IOException {
...
     }
 
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(Fields.DOCUMENT);
-        builder.field(Fields.MODIFIED, ingestDocument.isModified());
+        builder.field(Fields.MODIFIED, ingestDocument.isSourceModified());
         builder.field(Fields.INDEX, ingestDocument.getMetadata(INDEX));
         builder.field(Fields.TYPE, ingestDocument.getMetadata(TYPE));
         builder.field(Fields.ID, ingestDocument.getMetadata(ID));
```
Comment (javanna) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/WriteableIngestDocument.java]: we need to print out routing parent timestamp and ttl here too?

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/simulate/SimulatePipelineRequest.java [hunk 4]
```diff
@@ -91,5 +91,9 @@ public void writeTo(StreamOutput out) throws IOException {
         static final String INDEX = "_index";
         static final String TYPE = "_type";
         static final String ID = "_id";
+        static final String ROUTING = "_routing";
+        static final String PARENT = "_parent";
+        static final String TIMESTAMP = "_timestamp";
+        static final String TTL = "_ttl";
```
Comment (javanna) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/simulate/SimulatePipelineRequest.java]: shall we request the MetaData enum here too?

File: plugins/ingest/src/test/java/org/elasticsearch/ingest/processor/meta/MetaDataProcessorTests.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+        MetaDataProcessor processor = new MetaDataProcessor(templates);
+        IngestDocument ingestDocument = new IngestDocument("_index", "_type", "_id", Collections.singletonMap("field", "value"));
+        processor.execute(ingestDocument);
+
+        for (MetaData metaData : MetaData.values()) {
+            assertThat(ingestDocument.getMetadata(metaData), Matchers.equalTo("some value"));
+        }
+    }
+
```
Comment (javanna) [inline plugins/ingest/src/test/java/org/elasticsearch/ingest/processor/meta/MetaDataProcessorTests.java]: Is it worth to test what happens when the mustache expression evaluation breaks?

**Comment by javanna** [comment]
left a few comments, I like it!

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java (hunk 2 above)
Comment (martijnvg) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java]: right, I'll remove the metadata flag.

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/WriteableIngestDocument.java (hunk 3 above)
Comment (martijnvg) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/WriteableIngestDocument.java]: oops, I'll add it.

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java [hunk 6]
```diff
@@ -40,22 +43,47 @@ public PipelineExecutionService(PipelineStore store, ThreadPool threadPool) {
...
+            String index = indexRequest.index();
+            String type = indexRequest.type();
+            String id = indexRequest.id();
+            String routing = indexRequest.routing();
+            String parent = indexRequest.parent();
+            String timestamp = indexRequest.timestamp();
+            String ttl = null;
+            if (indexRequest.ttl() != -1) {
+                // At this point we don't know the original string ttl that was specified,
+                // so we return it as ms
```
Comment (javanna) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java]: I think this comment is outdated :)

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/simulate/SimulatePipelineRequest.java [hunk 7]
```diff
@@ -96,9 +97,13 @@ public void writeTo(StreamOutput out) throws IOException {
...
-        static final String INDEX = "_index";
-        static final String TYPE = "_type";
-        static final String ID = "_id";
+        static final String INDEX = MetaData.INDEX.getFieldName();
+        static final String TYPE = MetaData.TYPE.getFieldName();
+        static final String ID = MetaData.ID.getFieldName();
+        static final String ROUTING = MetaData.ROUTING.getFieldName();
+        static final String PARENT = MetaData.PARENT.getFieldName();
+        static final String TIMESTAMP = MetaData.TIMESTAMP.getFieldName();
+        static final String TTL = MetaData.TTL.getFieldName();
```
Comment (javanna) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/simulate/SimulatePipelineRequest.java]: sorry I thought we were using XContentBuilderString here, does it make sense to remove these 7 contants and use MetaData ones directly everywhere?

**Comment by javanna** [comment]
left two super minor comments, LGTM though, no need for another review

**Pull request merged**